### PR TITLE
Make solana-frozen-abi optional in all remaining crates

### LINKED ISSUE #1096: Make solana-frozen-abi optional everywhere
#### Problem
Afaict solana-frozen-abi is basically just for testing, but it's required as a prod dependency whenever it gets used.

#### Proposed Solution
Wherever frozen-abi gets used, put it behind a feature flag that is not activated by default

### PR DESCRIPTION
**Pull request by kevinheavey**
#### Summary of Changes
This could have been multiple extremely repetitive PRs but I think it's more efficient this way. It certainly would have taken me much longer to do the same find+replace over and over. 

Aside from the cfg and cfg_attr stuff, this PR also:
- Activates the `frozen-abi` feature in the scripts that are supposed to include ABI tests
- For any crate that has a frozen-abi feature, activates the frozen-abi feature on any relevant dependencies if it's activated in the crate itself
- Changes the dev-dependencies of solana-sdk to no longer activate frozen-abi in solana-program by default. This was added in #1222 

Fixes #1096

### CODE REVIEW

File: programs/sbf/Cargo.lock
Comment (joncinque) [inline programs/sbf/Cargo.lock:1]: Ah, now *these* are gains

File: programs/address-lookup-table/Cargo.toml [hunk 1]
```diff
@@ -15,8 +15,8 @@ bytemuck = { workspace = true }
 log = { workspace = true }
 num-derive = { workspace = true }
 num-traits = { workspace = true }
-solana-frozen-abi = { workspace = true }
-solana-frozen-abi-macro = { workspace = true }
+solana-frozen-abi = { workspace = true, optional = true }
+solana-frozen-abi-macro = { workspace = true, optional = true }
```
Comment (joncinque) [inline programs/address-lookup-table/Cargo.toml]: Does this crate need frozen abi at all?

File: sdk/Cargo.toml [hunk 2]
```diff
@@ -50,9 +53,12 @@ derivation-path = { workspace = true }
...
 ed25519-dalek = { workspace = true, optional = true }
 ed25519-dalek-bip32 = { workspace = true, optional = true }
-generic-array = { workspace = true, features = ["serde", "more_lengths"], optional = true }
+generic-array = { workspace = true, features = [
+  "serde",
+  "more_lengths",
+], optional = true }
 hmac = { workspace = true }
-itertools =  { workspace = true }
+itertools = { workspace = true }
```
Comment (joncinque) [inline sdk/Cargo.toml]: nit: Can you revert these changes? let's run a formatter over the toml files in a separate PR

File: sdk/Cargo.toml [hunk 3]
```diff
@@ -17,24 +17,27 @@ edition = { workspace = true }
...
+  "serde_json",
+  "ed25519-dalek",
+  "ed25519-dalek-bip32",
+  "libsecp256k1",
+  "sha3",
+  "digest",
 ]
-dev-context-only-utils = [
-  "qualifier_attr"
+dev-context-only-utils = ["qualifier_attr"]
```
Comment (joncinque) [inline sdk/Cargo.toml]: nit: Can you also revert these changes? Let's do it in a separate PR

File: core/Cargo.toml [hunk 4]
```diff
@@ -95,7 +95,9 @@ solana-program-runtime = { workspace = true }
 solana-runtime = { workspace = true, features = ["dev-context-only-utils"] }
 solana-sdk = { workspace = true, features = ["dev-context-only-utils"] }
 solana-stake-program = { workspace = true }
-solana-unified-scheduler-pool = { workspace = true, features = ["dev-context-only-utils"] }
+solana-unified-scheduler-pool = { workspace = true, features = [
+    "dev-context-only-utils",
+] }
```
Comment (joncinque) [inline core/Cargo.toml]: nit: can you revert this change so it can be done in a separate PR?

**Review by joncinque — COMMENTED** [review]
Looks really great overall! Thanks for taking this on. Just some small questions.

Btw, have you been measuring the effects of your PRs by any chance? I'm curious to find out what the gains add up to!

File: programs/address-lookup-table/Cargo.toml (hunk 1 above)
Comment (kevinheavey) [reply programs/address-lookup-table/Cargo.toml]: Looks like it doesn't! Removing this

**Review by joncinque — DISMISSED** [review]
Thanks for taking the time to clean this all up, your effort is much appreciated.

**Comment by kevinheavey** [comment]
> Btw, have you been measuring the effects of your PRs by any chance?

Ran a sloppy experiment on my laptop and it says solana-program 1.18.13 takes 13% longer to build than this branch

File: runtime/src/serde_snapshot.rs [hunk 5]
```diff
@@ -827,8 +831,8 @@ impl<'a> Serialize for SerializableAccountsDb<'a> {
     }
 }
 
-#[cfg(RUSTC_WITH_SPECIALIZATION)]
-impl<'a> solana_frozen_abi::abi_example::IgnoreAsHelper for SerializableAccountsDb<'a> {}
+#[cfg(all(RUSTC_WITH_SPECIALIZATION, feature = "frozen-abi"))]
+impl<'a, C> solana_frozen_abi::abi_example::IgnoreAsHelper for SerializableAccountsDb<'a, C> {}
```
Comment (joncinque) [inline runtime/src/serde_snapshot.rs]: Was this change intentional? It doesn't look like this type has a generic param

**Review by joncinque — COMMENTED** [review]
One little fixup, then we can land this

File: runtime/src/serde_snapshot.rs (hunk 5 above)
Comment (kevinheavey) [reply runtime/src/serde_snapshot.rs]: ah, this struct moved file and I copied over an old line

**Review by joncinque — APPROVED** [review]
